### PR TITLE
Add support for icons in terra application layout nav items

### LINKED ISSUE #409: Add support for ability to add icons top navigation
# Feature Request

## Description
Today we have apps where clients can assign icons to the navigation items.  A request has arisen that we allow for the same to occur in the terra-application-layout top navigation to help ease the migration to the terra-application-layout.

## @ Mentions
@mhemesath @nickpith @neilpfeiffer @tbiethman @alex-bezek  @dkasper-was-taken  @mjhenkes @dev-cprice @JakeLaCombe

### PR DESCRIPTION
**Pull request by dev-cprice**
### Summary
Closes #409

* Add ability to add icons to the application layout primary navigation items.
* Ensure passivity when not providing icons

### Deployment Link
https://terra-framework-deploye-pr-553.herokuapp.com/#/tests/terra-application-layout/application-layout/application-layout-with-icons

cc
@alex-bezek @mhemesath @bjankord

### CODE REVIEW

File: packages/terra-application-links/src/terra-dev-site/test/application-links/ApplicationTabsWithIcons.test.jsx [hunk 1]
```diff
@@ -0,0 +1,22 @@
+import React from 'react';
+import { MemoryRouter } from 'react-router-dom';
+import ApplicationTabs from '../../../tabs/ApplicationTabs';
+import testLinksWithIconsConfig from '../common/testLinksWithIconsConfig';
+
+const demoheaderstyles = {
```
Comment (bjankord) [inline packages/terra-application-links/src/terra-dev-site/test/application-links/ApplicationTabsWithIcons.test.jsx]: You'll want to move these styles to a SCSS file and likely want to make the background themable. See this PR for more info: https://github.com/cerner/terra-framework/pull/550

File: packages/terra-application-links/src/terra-dev-site/doc/example/ApplicationTabsWithIconsExample.jsx [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import React from 'react';
+/* eslint-disable import/no-extraneous-dependencies, import/no-unresolved, import/extensions */
+import { ApplicationTabs } from 'terra-application-links/lib/ApplicationLinks';
+import ApplicationLinkConfig from 'terra-application-links/lib/terra-dev-site/doc/common/ApplicationLinksWithIconsConfig';
+/* eslint-enable import/no-extraneous-dependencies, import/no-unresolved, import/extensions */
+
+const demoheaderstyles = {
```
Comment (bjankord) [inline packages/terra-application-links/src/terra-dev-site/doc/example/ApplicationTabsWithIconsExample.jsx]: You'll want to move these styles to a SCSS file and likely want to make the background themable. See this PR for more info: https://github.com/cerner/terra-framework/pull/550

File: packages/terra-application-layout/src/header/ApplicationHeader.module.scss [hunk 3]
```diff
@@ -22,6 +22,10 @@
     }
   }
 
+  .application-navbar-with-icons {
+    height: var(--terra-application-layout-navbar-with-icons-height, 5.142857143rem);
```
Comment (bjankord) [inline packages/terra-application-layout/src/header/ApplicationHeader.module.scss]: I think you'll want to add some code for a different height when the layout is in a narrow "mobile" screen. Something like

```scss
@import '~terra-breakpoints/lib/media-queries';

.application-navbar-with-icons {
  height: var(--terra-application-layout-navbar-with-icons-mobile-height, 3.5714285714rem);

  @include terra-mq-large-up {
    height: var(--terra-application-layout-navbar-with-icons-height, 5.142857143rem);
  }
}
```

You'll need to add [terra-breakpoints](https://engineering.cerner.com/terra-ui/#/components/terra-breakpoints/breakpoints/about) as a dependency to terra-application-layout to pull in the media query mixin.

File: packages/terra-application-layout/src/terra-dev-site/doc/application-layout/Example.2.doc.jsx [hunk 4]
```diff
@@ -6,18 +6,28 @@ import ExampleDoc from '../../../../docs/example.md';
...
+
 const DocPage = () => (
   <DocTemplate
     packageName={name}
     readme={ExampleDoc}
     srcPath={`https://github.com/cerner/terra-framework/tree/master/packages/${name}`}
     examples={[
       {
-        example: <div style={{ height: '500px' }}><ExampleApplication /></div>,
+        example: <div style={style}><ExampleApplication /></div>,
```
Comment (tbiethman) [inline packages/terra-application-layout/src/terra-dev-site/doc/application-layout/Example.2.doc.jsx]: You could include `title` and/or `description` properties for these examples to describe what each example is representing.

**Comment by bjankord** [comment]
@neilpfeiffer / @mmalaker This is ready for UX / Functional review

**Comment by mmalaker** [comment]
Functional verification completed. The navigation items with icons display and function as expected.

**Comment by neilpfeiffer** [comment]
investigating a small styling-flash issue at the `terra-mq-large-up` breakpoint.

**Comment by dev-cprice** [comment]
@neilpfeiffer any updates on the issue you found? what exactly are you seeing?

**Comment by neilpfeiffer** [comment]
Update @dev-cprice - Talked with @bjankord and both agreed we can leave the cross-over styling flash for an enhancement fix later. (Found some other minor render sequence items we could pair with it if we need to fix in the layout, or we could focus effort and make sure it is fixed in the new layout currently being built.)

**Comment by neilpfeiffer** [comment]
+1 for initial release of new feature, intended only for a specified subset of consuming projects. 
 
Created enhancement issue to future solve for render-sequence flashes here: https://github.com/cerner/terra-framework/issues/586

**Pull request merged**